api/deviceflow.go

```diff
@@ -0,0 +1,37 @@
+package api
+
+import "github.com/infrahq/infra/internal/validate"
+
+type ApproveDeviceFlowRequest struct {
+	UserCode string `json:"userCode" example:"BDSD-HQMK"`
+}
+
+type DeviceFlowResponse struct {
+	DeviceCode          string `json:"deviceCode" example:"NGU4QWFiNjQ5YmQwNG3YTdmZMEyNzQ3YzQ1YSA" note:"a code that a device will use to exchange for an access key after device login is approved"`
+	VerificationURI     string `json:"verificationURI" example:"https://infrahq.com/device" note:"This is the URL the user needs to enter into their browser to start logging in"`
+	UserCode            string `json:"userCode" example:"BDSD-HQMK" note:"This is the text the user will enter at the Verification URI"`
+	ExpiresInSeconds    int16  `json:"expiresIn" example:"1800" note:"The number of seconds that this set of values is valid"`
+	PollIntervalSeconds int8   `json:"interval" example:"5" note:"the number of seconds the device should wait between polling to see if the user has finished logging in"`
+}
+
+type PollDeviceFlowRequest struct {
+	DeviceCode string `json:"deviceCode"`
+}
+
+func (pdfr *PollDeviceFlowRequest) ValidationRules() []validate.ValidationRule {
+	return []validate.ValidationRule{
+		validate.String("device_code", pdfr.DeviceCode, 38, 38, validate.AlphaNumeric),
+	}
+}
+
+type DevicePollResponse struct {
+	Status        string         `json:"status,omitempty" note:"can be one of pending, rejected, expired, confirmed"`
+	DeviceCode    string         `json:"deviceCode,omitempty" example:""`
+	LoginResponse *LoginResponse `json:"login,omitempty"`
+}
+
+func (adfr *ApproveDeviceFlowRequest) ValidationRules() []validate.ValidationRule {
+	return []validate.ValidationRule{
+		validate.String("user_code", adfr.UserCode, 8, 9, append(validate.DeviceFlowUserCode, validate.CharRange{Low: '-', High: '-'})),
```

Comment:
This `append` is not safe, it could introduce a very subtle bug. When this is called there's a good chance that it will modify the array backing `validate.DeviceFlowUserCode`. If any other code does a similar append with a different second argument it will modify the array, which modifies this slice as well. We'd end up with data races and also the wrong values in the slice.

The rule I generally follow is:

> if the first argument to append is a variable, the result of append should only be assigned to the same variable as the first argument to append

To fix this you can invert the arguments:

```go
append([]validate.CharRange{{Low: '-', High: '-'}}, validate.DeviceFlowUserCode...)
```